tests/system_tests/test_artifacts/test_wandb_artifacts_api.py

```diff
@@ -605,3 +607,105 @@ def test_artifact_multipart_download(user: str, api: Api):
     # Verify checksum
     downloaded_md5 = md5_file_hex(os.path.join(stored_folder, file_path))
     assert downloaded_md5 == md5_value
+
+
+def _assert_headers_in_requests(calls, custom_headers):
+    """Assert that custom headers are present in all storage requests.
+
+    Args:
+        calls: responses.calls list from responses library
+        custom_headers: dict of expected headers
+    """
+    storage_keywords = [
+        "wandb_manifest.json",  # Manifest file always ends with wandb_manifest.json
+        "wandb_artifacts",  # Artifact files are stored under wandb_artifacts/ prefix
+    ]
+    storage_calls = [
+        call
+        for call in calls
+        if any(kw in call.request.url.lower() for kw in storage_keywords)
+    ]
+    assert len(storage_calls) > 0, (
+        f"Expected storage requests, got none. All requests: {[c.request.url for c in calls]}"
+    )
+    has_manifest = False
+    for call in storage_calls:
+        if "wandb_manifest.json" in call.request.url:
+            has_manifest = True
+        for header_name, header_value in custom_headers.items():
+            assert header_name in call.request.headers, (
+                f"Missing {header_name} in {call.request.url}"
+            )
+            assert call.request.headers[header_name] == header_value
+    assert has_manifest, "Expected manifest request, got none"
+    return storage_calls
+
+
+def test_artifact_download_http_headers(user, api, monkeypatch, tmp_path):
+    """Test custom HTTP headers are included in full artifact download requests."""
+    custom_headers = {
+        "X-Custom-Header": "test-value",
+        "X-Another-Header": "another-value",
+    }
+    monkeypatch.setenv("WANDB_X_EXTRA_HTTP_HEADERS", json.dumps(custom_headers))
+
+    entity = user
+    project = "test-http-headers-full"
+    artifact_name = "test-headers-artifact"
+
+    test_file = tmp_path / "a.txt"
+    test_file.write_text("a")
+    test_file2 = tmp_path / "b.txt"
+    test_file2.write_text("b")
+
+    with wandb.init(entity=entity, project=project) as run:
+        art = wandb.Artifact(artifact_name, "dataset")
+        art.add_file(test_file)
+        art.add_file(test_file2)
+        run.log_artifact(art)
+        art.wait()
+
+    # Full artifact download with responses passthrough to capture requests
+    with responses.RequestsMock(assert_all_requests_are_fired=False) as rsps:
+        # Passthrough all HTTP methods (POST for GraphQL, GET for downloads)
+        rsps.add(responses.PassthroughResponse(responses.GET, re.compile(".*")))
+        rsps.add(responses.PassthroughResponse(responses.POST, re.compile(".*")))
```

Comment:
suggestion: you can just use `.add_passthru()` AFAICT.  See the docs on passing through real responses: https://github.com/getsentry/responses?tab=readme-ov-file#passing-through-real-requests

```suggestion
        rsps.add_passthru(re.compile(".*"))
```